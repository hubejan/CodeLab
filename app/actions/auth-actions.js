// @flow
import axios from 'axios';
import gitAuth from '../utils/github.settings';

export const LOGIN_SUCCESS = 'LOGIN_SUCCESS';
export const LOGIN_FAILURE = 'LOGIN_FAILURE';
export const LOGOUT = 'LOGOUT';

type authInfoType = {
  client_id: string,
  client_secret: string,
  code: string
};

const authorizeApp = (authInfo: authInfoType) => {
  const url = 'https://github.com/login/oauth/access_token';
  const config = {
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
      'Cache-Control': 'no-cache'
    }
  };
  return axios.post(url, authInfo, config);
};

const getUsername = (token) => {
  const url = 'https://api.github.com/user';
  const config = {
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
      'Cache-Control': 'no-cache',
      Authorization: `token ${token}`
    }
  };
  return axios.get(url, config)
    .then(response => response.data.name); // optionally data.login
};

const revokeAuth = (token: string) => {
  const url = `https://api.github.com/applications/${gitAuth.client_id}/grants/${token}`;
  const config = {
    auth: {
      username: gitAuth.client_id,
      password: gitAuth.client_secret
    }
  };
  return axios.delete(url, config);
};

export const storageLogin = (token: string, username: string) => ({
  type: LOGIN_SUCCESS,
  token,
  username
});

export const teacherLogin = (authCode: string) => (dispatch: *) => {
  const authInfo = {
    client_id: gitAuth.client_id,
    client_secret: gitAuth.client_secret,
    code: authCode
  };

  return authorizeApp(authInfo)
    .then(response => {
      if (!response.data.access_token) {
        const error = new Error('failed authorization');
        error.failure = response.data;
        throw error;
      }
      return response.data.access_token;
    })
    .then(token => Promise.all([token, getUsername(token)]))
    .then(([token, username]) => {
      localStorage.setItem('token', token);
      localStorage.setItem('username', username);
      return dispatch({ type: LOGIN_SUCCESS, token, username });
    })
    .catch(error => {
      console.error(error);
      dispatch({ type: LOGIN_FAILURE, error: error.failure });
    });
};

// future: come back and learn how to do dispatch/getState flow types
export const teacherLogout = () => (dispatch: *, getState: *) =>
  revokeAuth(getState().auth.token)
    .then(() => {
      localStorage.clear();
      return dispatch({ type: LOGOUT });
    })
    .catch(error => {
      console.error(error);
    });
