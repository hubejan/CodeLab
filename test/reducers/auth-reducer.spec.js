import authReducer from '../../app/reducers/auth-reducer';
import { LOGIN_SUCCESS, LOGIN_FAILURE, LOGOUT } from '../../app/actions/auth-actions';

describe('auth reducer', () => {
  const loggedOutState = {
    token: null,
    error: {},
    loggedIn: false
  };

  const loggedInState = {
    token: 'anAccessToken',
    error: {},
    loggedIn: true,
    username: 'Test Name'
  };

  it('should handle initial state', () => {
    expect(authReducer(undefined, {})).toMatchSnapshot();
  });

  it('should handle LOGIN_SUCCESS', () => {
    const successAction = { type: LOGIN_SUCCESS, token: 'anAccessToken', username: 'Test Name' };
    expect(authReducer(loggedOutState, successAction)).toMatchSnapshot();
  });

  it('should handle LOGIN_FAILURE', () => {
    const error = { error: 'bad validation', error_description: 'incorrect code', error_uri: '/error' };
    const failureAction = { type: LOGIN_FAILURE, error };
    expect(authReducer(loggedOutState, failureAction)).toMatchSnapshot();
  });

  it('should handle LOGOUT', () => {
    expect(authReducer(loggedInState, { type: LOGOUT })).toMatchSnapshot();
  });
});
