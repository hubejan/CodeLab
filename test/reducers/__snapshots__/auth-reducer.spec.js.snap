// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth reducer should handle LOGIN_FAILURE 1`] = `
Object {
  "error": Object {
    "error": "bad validation",
    "error_description": "incorrect code",
    "error_uri": "/error",
  },
  "loggedIn": false,
  "token": null,
}
`;

exports[`auth reducer should handle LOGIN_SUCCESS 1`] = `
Object {
  "error": Object {},
  "loggedIn": true,
  "token": "anAccessToken",
}
`;

exports[`auth reducer should handle LOGOUT 1`] = `
Object {
  "error": Object {},
  "loggedIn": false,
  "token": null,
}
`;

exports[`auth reducer should handle initial state 1`] = `
Object {
  "error": Object {},
  "loggedIn": false,
  "token": null,
}
`;
